Mathlib/NumberTheory/FLT/Basic.lean

```diff
@@ -153,6 +153,74 @@ lemma fermatLastTheoremFor_iff_int {n : ℕ} : FermatLastTheoremFor n ↔ Fermat
 lemma fermatLastTheoremFor_iff_rat {n : ℕ} : FermatLastTheoremFor n ↔ FermatLastTheoremWith ℚ n :=
   (fermatLastTheoremWith_nat_int_rat_tfae n).out 0 2
 
+/--
+A relaxed variant of Fermat's Last Theorem over a given commutative semiring with a specific
+exponent, allowing nonzero solutions of units and their common multiples.
+
+1. The variant `FermatLastTheoremWith' α` is weaker than `FermatLastTheoremWith α` in general.
+   In particular, it holds trivially for `[Field α]`.
+2. This variant is equivalent to the original `FermatLastTheoremWith α` for `α = ℕ` or `ℤ`.
+   In general, they are equivalent if there is no solutions of units to the Fermat equation.
+3. For a polynomial ring `α = k[X]`, the original `FermatLastTheoremWith α` is false but the weaker
+   variant `FermatLastTheoremWith' α` is true. This polynomial variant of Fermat's Last Theorem
+   can be shown elementarily using Mason--Stothers theorem.
+-/
+def FermatLastTheoremWith' (α : Type*) [CommSemiring α] (n : ℕ) : Prop :=
+  ∀ a b c : α, a ≠ 0 → b ≠ 0 → c ≠ 0 → a ^ n + b ^ n = c ^ n →
+    ∃ d a' b' c', (a = a' * d ∧ b = b' * d ∧ c = c' * d) ∧ (IsUnit a' ∧ IsUnit b' ∧ IsUnit c')
+
+lemma fermatLastTheoremWith'_of_fermatLastTheoremWith {α : Type*} [CommSemiring α] {n : ℕ}
+    (h : FermatLastTheoremWith α n) : FermatLastTheoremWith' α n :=
```

Comment:
Probably this should be named to allow dot notation: `FermatLastTheoremWith.fermatLastTheoremWith'`